README.md
# auto-patch [![Build Status](https://secure.travis-ci.org/bflad/chef-auto-patch.png?branch=master)](http://travis-ci.org/bflad/chef-auto-patch)

## Description

Chef Cookbook for automatically patching nodes on a specific schedule (weekday,
hour, and minute). Handles weekly or monthly patching routines with or without
node splay for large environments. Can automatically prepare node for true patch
window by cleaning and pre-downloading packages, which speeds up patch process
and can help guarantee meeting patching timeframes.

If you're looking for daily upgrades on Ubuntu, check out the
[unattended-upgrades package](https://help.ubuntu.com/12.04/serverguide/automatic-updates.html)
and a [corresponding cookbook](https://github.com/firstbanco/chef-unattended-upgrades).
(/ht [@hectcastro](https://github.com/hectcastro) for this information).

## Requirements

### Platforms

* RedHat 6.3 (Santiago)
* Ubuntu 12.04 (Precise)

### Cookbooks

* cron

## Attributes

* `node["auto-patch"]["disable"]` - defaults to false
* `node["auto-patch"]["hour"]` - defaults to 3
* `node["auto-patch"]["minute"]` - defaults to 0
* `node["auto-patch"]["monthly"]` - auto patching occurs once a month on the
  corresponding textual week number ("first","second",etc) and weekday
  ("monday","tuesday",etc), overridden by `node["auto-patch"]["weekly"]`,
  defaults to "first sunday"
* `node["auto-patch"]["reboot"]` - reboot automatically after patching, defaults
  to true
* `node["auto-patch"]["splay"]` - seconds of random delay before beginning,
  defaults to 0
* `node["auto-patch"]["weekly"]` - auto patching occurs once a week on the
  corresponding textual weekday ("monday","tuesday",etc), overrides
  `node["auto-patch"]["monthly"]`, defaults to nil
* `node["auto-patch"]["now"]`, patches the system immediately, default to false

### Patch Preparation

* `node["auto-patch"]["prep"]["clean"]` - cleans updater cache files, defaults
  to true
* `node["auto-patch"]["prep"]["disable"]` - defaults to true
* `node["auto-patch"]["prep"]["hour"]` - defaults to 2
* `node["auto-patch"]["prep"]["minute"]` - defaults to 0
* `node["auto-patch"]["prep"]["monthly"]` - auto patching prep occurs once a
  month on the corresponding textual week number ("first","second",etc) and
  weekday ("monday","tuesday",etc), overridden by
  `node["auto-patch"]["prep"]["weekly"]`, defaults to "first sunday"
* `node["auto-patch"]["prep"]["splay"]` - seconds of random delay before
  beginning, defaults to 1800
* `node["auto-patch"]["prep"]["weekly"]` - auto patching prep occurs once a week
  on the corresponding textual weekday ("monday","tuesday",etc), overrides
  `node["auto-patch"]["prep"]["monthly"]`, defaults to nil
* `node["auto-patch"]["prep"]["update_updater"]` - updates apt or yum before
  actual patching occurs, defaults to true

## Recipe

* `recipe[auto-patch]` configures automatic patching

## Usage

* Change any attributes to fit your patching cycle
* If using auto patch preparation, ensure it starts before auto patch (remember
  any splay!)
* Add `recipe[auto-patch]` to your node's run list

### Update Once

Set `node["auto-patch"]["once"]` attribute to true will set the system to update
and reboot by default a single instance at the date and time indicated. 

### Weekly automatic patching

Just use the `node["auto-patch"]["weekly"]` attribute to override the monthly
setting. Don't forget to add appropriate `node["auto-patch"]["prep"]["weekly"]`
if you're using automatic patch preparation.

### Automatic patching of large numbers of nodes

If you're auto patching many nodes at once, you have two attributes you can
optionally add to prevent denial of service against your network, update
server(s), and resources:
* Adding `node["auto-patch"]["splay"]`
* Setting `node["auto-patch"]["prep"]["disable"]` to false (along with setting
  associated prep attributes to meet your patch cycle)

### Using roles to specify auto patching groups

If you'd like to specify groups of nodes for auto patching, you can setup roles.

Say you want to auto patch some nodes at 8am and some at 10pm on your monthly
"patch day" of the fourth Wednesday every month.

If you have a base role (you do, right?), you can save duplicating attributes
and specify some base information first:

    "auto-patch" => {
      "monthly" => "fourth wednesday",
      "prep" => {
        "disable" => false,
        "hour" => 7,
        "monthly" => "fourth wednesday"
      }
    }

Example role that then could be added to 8am nodes:

    name "auto-patch-0800"
    description "Role for automatically patching nodes at 8am on patch day."
    default_attributes(
      "auto-patch" => {
        "hour" => 8
      }
    )
    run_list(
      "recipe[auto-patch]"
    )

Example role that then could be added to 10pm nodes:

    name "auto-patch-2200"
    description "Role for automatically patching nodes at 10pm on patch day."
    default_attributes(
      "auto-patch" => {
        "hour" => 22
      }
    )
    run_list(
      "recipe[auto-patch]"
    )

### Disabling automatic patching

* Specify `node["auto-patch"]["disable"]` to true
* If necessary, specify `node["auto-patch"]["prep"]["disable"]` to true
* Run chef-client on your node

## Recipe

* `recipe[auto-patch::update-now]` performs patching

### Update now

* Specify `node["auto-patch"]["now"]` to true - node will install all updates during each chef run

## Testing and Development

This repo includes a .kitchen.yml file which leverages [Test Kitchen](http://kitchen.ci/) and [Vagrant](http://docs.vagrantup.com/v2/cli/index.html) to test and verify.  

Test Kitchen is already included as part of [ChefDK](https://downloads.chef.io/chef-dk/).  To start testing run:

    kitchen converge

Only the system has finished converging you can log into it by using the command

    kitchen login

All rspec and integration tests are performed and the system can be manually verified.

## Contributing

Please use standard Github issues/pull requests.

## License and Author
      
Author:: Brian Flad (<bflad417@gmail.com>)
         Derek Groh (<derekgroh@gmail.com>)

Copyright:: 2012-2013

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## 0.2.0

* Add features update_once and update_now

## 0.1.10

* Bugfix: #2 check if constants are already defined

## v0.1.9 ##

* RHEL 5 support with correct package name

## v0.1.8 ##

* Fixed day of month logic for later week specifications

## v0.1.7 ##

* Removed cookbook attribute from cron_d resources since was fixed in [cron@1.2.2](https://github.com/opscode-cookbooks/cron/compare/1.2.0...1.2.2)

## v0.1.6

* Fixed typo in default recipe for platform_family detection

## v0.1.5

* Fixes in default recipe for WARN: Setting attributes without specifying a precedence is deprecated

## v0.1.4

* Defaulted auto-patch splay to 0 since setting is only for large numbers of
  nodes

## v0.1.3

* Refactored next_monthly_*_date functions into next_monthly_date

## v0.1.2

* Convert unnecessary String attributes to Fixnum

## v0.1.1

* Only install yum-plugin-downloadonly if auto-patch prep is enabled.

## v0.1.0

* Initial release with tested Red Hat 6.3 support

## v0.0.1

* Initial cookbook building and testing.


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: auto-patch
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'auto-patch::default' do
  context 'When all attributes are default, on an unspecified platform' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['auto-patch']['prep']['disable'] = true
        node.set['auto-patch']['disable'] = true
      end.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run.to_not raise_error }
    end

    it 'includes the `cron` recipe' do
      expect(chef_run).to include_recipe('cron::default')
    end

    it 'deletes the template `/usr/local/sbin/auto-patch-prep` if node[\'auto-patch\'][\'prep\'][\'disable\'] is true' do
      expect(chef_run).to delete_template('/usr/local/sbin/auto-patch-prep').with(
        :source => 'auto-patch-prep.sh.erb',
        :owner => 'root',
        :group => 'root',
        :mode => '0700'
      )
    end

    it 'deletes the cron `auto-patch-prep` if node[\'auto-patch\'][\'prep\'][\'disable\'] is true' do
      expect(chef_run).to delete_cron_d('auto-patch-prep').with(
        #:hour => '',
        #:minute => '',
        #:weekday => '',
        #:day => '',
        #:month => '',
        :command => '/usr/local/sbin/auto-patch-prep'
      )
    end

    it 'deletes the template `/usr/local/sbin/auto-patch` if node[\'auto-patch\'][\'disable\']' do
      expect(chef_run).to delete_template('/usr/local/sbin/auto-patch').with(
        :source => 'auto-patch.sh.erb',
        :owner => 'root',
        :group => 'root',
        :mode => '0700'
      )
    end

    it 'deletes the cron `auto-patch` if node[\'auto-patch\'][\'disable\'] is true' do
      expect(chef_run).to delete_cron_d('auto-patch').with(
        #:hour => '',
        #:minute => '',
        #:weekday => '',
        #:day => '',
        #:month => '',
        :command => '/usr/local/sbin/auto-patch'
      )
    end
  end
end

describe 'auto-patch::default' do
  context 'When `node[\'auto-patch\'][\'now\']` is true, on an unspecified platform' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['auto-patch']['now'] = true
      end.converge(described_recipe)
    end

    it 'includes the `auto-patch::update_now` recipe' do
      expect(chef_run.node['auto-patch']['now']).to eq(true)
      expect(chef_run).to include_recipe('auto-patch::update_now')
    end
  end
end

describe 'auto-patch::default' do
  context 'when `node[\'auto-patch\'][\'once\']` is true, on an unspecified platform' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new do |node|
        node.set['auto-patch']['once'] = true
      end.converge(described_recipe)
    end

    it 'includes the `auto-patch::update_once` recipe' do
      expect(chef_run.node['auto-patch']['once']).to eq(true)
      expect(chef_run).to include_recipe('auto-patch::update_once')
    end
  end
end

describe 'auto-patch::default' do
  context 'When all attributes are default, on a `rhel < 6.0` platform' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'rhel', version: '<6.0')
      runner.converge(described_recipe)
    end

    it 'installs a package `yum-downloadonly' do
      expect { chef_run.to install_package('yum-downloadonly') }
    end
  end
end

describe 'auto-patch::default' do
  context 'When all attributes are default, on a `rhel > 6.0` platform' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'rhel', version: '>6.0')
      runner.converge(described_recipe)
    end

    it 'installs a package `yum-plugin-downloadonly' do
      expect { chef_run.to install_package('yum-plugin-downloadonly') }
    end
  end
end


spec/unit/recipes/update_now_spec.rb
#
# Cookbook Name:: chef-auto-patch
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'auto-patch::update_now' do
  context 'when on CentOS' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'rhel')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run.to_not raise_error }
    end

    it 'runs an execute `Yum update`' do
      expect do
        chef_run.to run_execute('Yum update').with(
          :command => 'yum -y update'
        )
      end
    end

    it 'writes a log `rhel`' do
      expect do
        chef_run.to write_log('rhel').with(
          :level => ':info',
          :message => 'yum update performed'
        )
      end
    end
  end

  context 'when on Ubuntu' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run.to_not raise_error }
    end

    it 'runs an execute `apt-get update`' do
      expect do
        chef_run.to run_execute('apt-get update').with(
          :command => 'apt-get -q update'
        )
      end
    end

    it 'writes a log `debian`' do
      expect do
        chef_run.to write_log('debian').with(
          :level => ':info',
          :message => 'apt-get update performed'
        )
      end
    end
  end
end


recipes/update_now.rb
#
# Cookbook Name:: chef-auto-patch
# Recipe:: update_now
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

if node['auto-patch']['now']
  case node['platform_family']
  when 'rhel'
    execute 'Yum update' do
      command 'yum -y update'
    end
    log 'rhel' do
      level :info
      message 'yum update performed'
    end
  when 'debian'
    execute 'apt-get update' do
      command 'apt-get -q update'
    end
    log 'debian' do
      level :info
      message 'apt-get update performed'
    end
  else
    log 'update-now failed - platform_family' do
      level :warn
      message '***** WARNING platform_family not recognized - update failed *****'
    end
  end
else
  log 'update-now failed' do
    level :warn
    message "***** WARNING node['auto-patch']['now'] set to false, system not patched *****"
  end
end


recipes/update_once.rb
#
# Cookbook Name:: auto-patch
# Recipe:: update_once
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

if node['auto-patch']['once']
  case node['platform_family']
  when 'rhel'
    package 'at' do
      action :install
    end
    service 'atd' do
      action :start
    end
    execute 'at yum update' do
      command 'echo "sudo yum -y update" | at now'
    end
    log 'rhel' do
      level :info
      message 'echo "sudo yum -y update" | at now executed'
    end
  when 'ubuntu'
    execute 'at apt-get -q update' do
      command 'echo "sudo apt-get -q update" | at now'
    end
    log 'debian' do
      level :info
      message 'echo "sudo apt-get -q update" | at now'
    end
  else
    log 'update-once failed - platform_family' do
      level :warn
      message '***** WARNING platform_family not recognized - update failed *****'
    end
  end
else
  log 'update-once failed' do
    level :warn
    message "***** WARNING node['auto-patch']['once'] set to false, system not patched *****"
  end
end


libraries/auto_patch.rb
# Cookbook Name:: auto-patch
# Library:: AutoPatch
#
# Copyright 2012
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'date'

# auto-patch helper logic
# Chef class
class Chef
  # Chef::Recipe class
  class Recipe
    # Chef::Recipe::AutoPatch class
    class AutoPatch
      WEEKS = %w(first second third fourth).freeze unless defined?(WEEKS)
      WEEKDAYS = %w(sunday monday tuesday wednesday thursday friday saturday).freeze unless defined?(WEEKDAYS)

      def self.monthly_date(year, month, monthly_specifier)
        Date.new(year, month, monthly_day(year, month, monthly_specifier))
      end

      def self.monthly_day(year, month, monthly_specifier)
        week, weekly_specifier = monthly_specifier.split(' ')
        week.downcase!
        weekly_specifier.downcase!
        Chef::Application.fatal!('Unknown week specified.') unless WEEKS.include?(week)

        first_day_occurance = 1
        while weekday(weekly_specifier) != Date.new(year, month, first_day_occurance).wday
          first_day_occurance += 1
        end
        first_day_occurance + (WEEKS.index(week) * 7)
      end

      def self.next_monthly_date(monthly_specifier, hour, minute)
        current_time = Time.now
        current_patch_time = Time.new(
          current_time.year,
          current_time.month,
          monthly_day(current_time.year, current_time.month, monthly_specifier),
          hour,
          minute
        )

        if current_time > current_patch_time
          if current_time.month == 12
            date = monthly_date(current_time.year + 1, 1, monthly_specifier)
          else
            date = monthly_date(current_time.year, current_time.month + 1, monthly_specifier)
          end
        else
          date = current_patch_time
        end

        date
      end

      def self.weekday(weekly_specifier)
        Chef::Application.fatal!('Unknown weekday specified.') unless WEEKDAYS.include?(weekly_specifier)
        WEEKDAYS.index(weekly_specifier)
      end
    end
  end
end


recipes/default.rb
# Cookbook Name:: auto-patch
# Recipe:: default
#
# Copyright 2012
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

include_recipe 'cron'
include_recipe 'auto-patch::update_now' if node['auto-patch']['now'] == true
include_recipe 'auto-patch::update_once' if node['auto-patch']['once'] == true

unless node['auto-patch']['prep']['disable']
  if node['platform_family'] == 'rhel'
    if node['platform_version'].to_f >= 6.0
      package 'yum-plugin-downloadonly'
    else
      package 'yum-downloadonly'
    end
  end

  if node['auto-patch']['prep']['weekly']
    node.set['auto-patch']['prep']['day'] = '*'
    node.set['auto-patch']['prep']['month'] = '*'
    node.set['auto-patch']['prep']['weekday'] = AutoPatch.weekday(node['auto-patch']['prep']['weekly'])
    Chef::Log.info("Auto patch prep scheduled weekly on #{node['auto-patch']['weekly']} at #{node['auto-patch']['prep']['hour']}:#{node['auto-patch']['prep']['minute']}")
  elsif node['auto-patch']['prep']['monthly']
    next_date = AutoPatch.next_monthly_date(
      node['auto-patch']['prep']['monthly'],
      node['auto-patch']['prep']['hour'],
      node['auto-patch']['prep']['minute'])
    node.set['auto-patch']['prep']['day'] = next_date.day
    node.set['auto-patch']['prep']['month'] = next_date.month
    node.set['auto-patch']['prep']['weekday'] = '*'
    Chef::Log.info("Auto patch prep scheduled for #{next_date.strftime('%Y-%m-%d')} at #{node['auto-patch']['prep']['hour']}:#{node['auto-patch']['prep']['minute']}")
  else
    Chef::Application.fatal!('Missing auto-patch prep monthly or weekly specification.')
  end
end

template '/usr/local/sbin/auto-patch-prep' do
  source 'auto-patch-prep.sh.erb'
  owner 'root'
  group 'root'
  mode '0700'
  action :delete if node['auto-patch']['prep']['disable']
end

cron_d 'auto-patch-prep' do
  hour node['auto-patch']['prep']['hour']
  minute node['auto-patch']['prep']['minute']
  weekday node['auto-patch']['prep']['weekday']
  day node['auto-patch']['prep']['day']
  month node['auto-patch']['prep']['month']
  command '/usr/local/sbin/auto-patch-prep'
  action :delete if node['auto-patch']['prep']['disable']
end

unless node['auto-patch']['disable']
  if node['auto-patch']['weekly']
    node.set['auto-patch']['day'] = '*'
    node.set['auto-patch']['month'] = '*'
    node.set['auto-patch']['weekday'] = AutoPatch.weekday(node['auto-patch']['weekly'])
    Chef::Log.info("Auto patch scheduled weekly on #{node['auto-patch']['weekly']} at #{node['auto-patch']['hour']}:#{node['auto-patch']['minute']}")
  elsif node['auto-patch']['monthly']
    next_date = AutoPatch.next_monthly_date(
      node['auto-patch']['monthly'],
      node['auto-patch']['hour'],
      node['auto-patch']['minute'])
    node.set['auto-patch']['day'] = next_date.day
    node.set['auto-patch']['month'] = next_date.month
    node.set['auto-patch']['weekday'] = '*'
    Chef::Log.info("Auto patch scheduled for #{next_date.strftime('%Y-%m-%d')} at #{node['auto-patch']['hour']}:#{node['auto-patch']['minute']}")
  else
    Chef::Application.fatal!('Missing auto-patch monthly or weekly specification.')
  end
end

template '/usr/local/sbin/auto-patch' do
  source 'auto-patch.sh.erb'
  owner 'root'
  group 'root'
  mode '0700'
  action :delete if node['auto-patch']['disable']
end

cron_d 'auto-patch' do
  hour node['auto-patch']['hour']
  minute node['auto-patch']['minute']
  weekday node['auto-patch']['weekday']
  day node['auto-patch']['day']
  month node['auto-patch']['month']
  command '/usr/local/sbin/auto-patch'
  action :delete if node['auto-patch']['disable']
end
